pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -244,47 +247,53 @@ func (gce *GCECloud) targetPoolURL(name, region string) string {
 	return fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/regions/%s/targetPools/%s", gce.projectID, region, name)
 }
 
-func waitForOp(op *compute.Operation, getOperation func() (*compute.Operation, error)) error {
-	pollOp := op
-	consecPollFails := 0
-	for pollOp.Status != "DONE" {
-		var err error
-		time.Sleep(3 * time.Second)
-		pollOp, err = getOperation()
-		if err != nil {
-			if consecPollFails == 2 {
-				// Only bail if we've seen 3 consecutive polling errors.
-				return err
-			}
-			consecPollFails++
-		} else {
-			consecPollFails = 0
-		}
+func waitForOp(op *compute.Operation, getOperation func(operationName string) (*compute.Operation, error)) error {
+	if op == nil {
+		return fmt.Errorf("operation must not be nil")
 	}
-	if pollOp.Error != nil && len(pollOp.Error.Errors) > 0 {
-		return &googleapi.Error{
-			Code:    int(pollOp.HttpErrorStatusCode),
-			Message: pollOp.Error.Errors[0].Message,
+
+	if opIsDone(op) {
+		return getErrorFromCompletedOp(op)
+	}
+
+	return wait.Poll(operationPollInterval, operationPollTimeoutDuration, func() (bool, error) {
+		pollOp, err := getOperation(op.Name)
+		glog.Warningf("GCE poll operation failed: %v", err)
+		return opIsDone(pollOp), getErrorFromCompletedOp(pollOp)
+	})
+}
+
+func opIsDone(op *compute.Operation) bool {
+	return op != nil && op.Status == "DONE"
+}
+
+func getErrorFromCompletedOp(op *compute.Operation) error {
+	if op != nil && op.Error != nil && len(op.Error.Errors) > 0 {
+		err := &googleapi.Error{
+			Code:    int(op.HttpErrorStatusCode),
+			Message: op.Error.Errors[0].Message,
 		}
+		glog.Errorf("GCE operation failed: %v", err)
+		return err
 	}
-	return nil
 
+	return nil
 }
 
 func (gce *GCECloud) waitForGlobalOp(op *compute.Operation) error {
-	return waitForOp(op, func() (*compute.Operation, error) {
+	return waitForOp(op, func(operationName string) (*compute.Operation, error) {
 		return gce.service.GlobalOperations.Get(gce.projectID, op.Name).Do()
```

Comment:
This will now crash here if operation is nil. If you make the func `func(op *compute.Operation) (*compute.Operation, error)`, then waitForOp can be the only one that needs to check for nil. Then you'd also need to handle errors a little bit differently above (you would only want to assign the return value of `getOperation` to `pollOp` if `err` was nil).

Or you could do it the opposite way: Check for nil in each of the three `waitFor***Op()` functions. Then have waitForOp just be:
```
func waitForOp(getOperation func() (*compute.Operation, error)) error {
	op, err := getOperation()
	...
}
```